File: typescript/tests/theory_of_mind.test.ts
// theory_of_mind.test.ts
//
// Verifies BeliefTrackerTheoryOfMind against the C# reference
// (HerJarvisRealImplementations.cs #10). The headline guarantees:
//   * the greedy claim group stops only at . ; ! ? (a run of clauses with no
//     delimiter is ONE claim),
//   * verb weighting (believe=1.0, others=0.7) with recency decay,
//   * `likelyBeliefJson` is byte-identical to
//     JsonSerializer.Serialize(Dictionary<string,double>) — including STJ's
//     aggressive default escaping and shortest-round-trip doubles.
// The expected JSON/confidence strings were captured from a .NET 10 probe that
// runs the identical algorithm.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  BeliefTrackerTheoryOfMind,
  stjEscape,
  stjDouble,
  stjSerializeDoubleMap,
} from '../src/companion/reasoning/index';

const tom = new BeliefTrackerTheoryOfMind();

describe('BeliefTrackerTheoryOfMind — belief extraction + wire format', () => {
  it('reproduces the C# JSON for a three-clause history (delimiter-separated)', async () => {
    const est = await tom.estimateAsync(
      'peer',
      'She thinks the plan is risky. He believes it will work. They want more time.',
    );
    assert.equal(est.targetIdentifier, 'peer');
    assert.equal(
      est.likelyBeliefJson,
      '{"thinks:the plan is risky":0.7,"believes:it will work":0.9090909090909091,"want:more time":0.5833333333333334}',
    );
    // conf = min(1, (0.7 + 0.9090909090909091 + 0.5833333333333334) / 5)
    assert.ok(Math.abs(est.confidence - 0.4384848484848485) < 1e-15);
  });

  it('the claim group is greedy up to . ; ! ? — no delimiter means one claim', async () => {
    // "Alex believes X and Bob THINKS x" has no . ; ! ? so the whole tail is the
    // claim; the second verb is swallowed, not matched. Weight 1.0, decay 1.0.
    const est = await tom.estimateAsync('peer', 'Alex believes X and Bob THINKS x');
    assert.equal(est.likelyBeliefJson, '{"believes:X and Bob THINKS x":1}');
    assert.ok(Math.abs(est.confidence - 0.2) < 1e-15);
  });

  it('splits on ! and ; as well as .', async () => {
    const est = await tom.estimateAsync(
      'peer',
      'He hopes for peace! She fears the dark; they think loudly?',
    );
    assert.equal(
      est.likelyBeliefJson,
      '{"hopes:for peace":0.7,"fears:the dark":0.6363636363636364,"think:loudly":0.5833333333333334}',
    );
  });

  it('empty belief set → {} and confidence 0', async () => {
    const est = await tom.estimateAsync('peer', 'nothing to see here, just a plain sentence');
    assert.equal(est.likelyBeliefJson, '{}');
    assert.equal(est.confidence, 0);
  });

  it('accumulates a repeated (verb,claim) key rather than duplicating it', async () => {
    // "wants coffee and wants tea" — one match, claim "coffee and wants tea".
    const est = await tom.estimateAsync('peer', 'user wants coffee and wants tea');
    assert.equal(est.likelyBeliefJson, '{"wants:coffee and wants tea":0.7}');
  });

  it('confidence clamps to 1.0 for many strong beliefs', async () => {
    const history =
      'x believes a. x believes b. x believes c. x believes d. x believes e. x believes f.';
    const est = await tom.estimateAsync('peer', history);
    assert.ok(est.confidence <= 1.0);
    assert.equal(est.confidence, Math.min(1.0, est.confidence)); // never exceeds 1
  });

  it('throws on blank target and null history', async () => {
    await assert.rejects(() => tom.estimateAsync('  ', 'x thinks y'), /target required/);
    await assert.rejects(
      // @ts-expect-error deliberately passing null to hit the guard
      () => tom.estimateAsync('peer', null),
      /interactionHistoryJson required/,
    );
  });
});

describe('System.Text.Json-faithful serialisation helpers', () => {
  it('escapes exactly like STJ default (aggressive: < > & + \' " ` and non-ASCII)', () => {
    assert.equal(stjEscape('angle <b> & plus+'), 'angle \\u003Cb\\u003E \\u0026 plus\\u002B');
    assert.equal(stjEscape("apostrophe ' its"), 'apostrophe \\u0027 its');
    // '"' becomes " (NOT \"), backslash stays \\, forward slash is bare.
    assert.equal(stjEscape('a"b\\c/d'), 'a\\u0022b\\\\c/d');
    // non-ASCII -> uppercase \uXXXX
    assert.equal(stjEscape('café'), 'caf\\u00E9');
    assert.equal(stjEscape('日'), '\\u65E5');
    // short escapes
    assert.equal(stjEscape('\t\n\r'), '\\t\\n\\r');
    // backtick and DEL
    assert.equal(stjEscape('`'), '\\u0060');
    assert.equal(stjEscape('\x7f'), '\\u007F');
  });

  it('formats doubles as shortest round-trip in normal range', () => {
    assert.equal(stjDouble(1.0), '1');
    assert.equal(stjDouble(0.7), '0.7');
    assert.equal(stjDouble(1.0 / 1.1), '0.9090909090909091');
    assert.equal(stjDouble(0), '0');
    assert.equal(stjDouble(-1), '-1');
  });

  it('formats scientific notation the .NET way (uppercase E, padded exponent)', () => {
    assert.equal(stjDouble(1e21), '1E+21');
    assert.equal(stjDouble(1e-7), '1E-07');
  });

  it('serialises an empty map as {} and preserves insertion order', () => {
    assert.equal(stjSerializeDoubleMap(new Map()), '{}');
    const m = new Map<string, number>([
      ['b', 0.5],
      ['a', 1.0],
    ]);
    assert.equal(stjSerializeDoubleMap(m), '{"b":0.5,"a":1}');
  });
});

File: rust/tests/theory_of_mind_test.rs
//! theory_of_mind_test.rs
//!
//! Verifies the companion theory of mind: BeliefTrackerTheoryOfMind, which does
//! bag-of-belief inference with confidence decay over an interaction history.
//! Mirrors the C# ITheoryOfMind.EstimateAsync behaviour (belief-verb scanning,
//! believe-weighting, decay, JSON serialisation, confidence).

use circle_ai::companion::theory_of_mind::{
    BeliefTrackerTheoryOfMind, ITheoryOfMind,
};
use serde_json::Value;

fn beliefs_of(estimate: &circle_ai::companion::theory_of_mind::OtherMindEstimate) -> Value {
    serde_json::from_str(&estimate.likely_belief_json).expect("valid belief json")
}

#[test]
fn no_belief_verbs_yields_empty_map_and_zero_confidence() {
    let tom = BeliefTrackerTheoryOfMind::new();
    let e = tom.estimate("alice", "the weather report for today");
    assert_eq!(e.target_identifier, "alice");
    assert_eq!(e.confidence, 0.0);
    let map = beliefs_of(&e);
    assert!(map.as_object().unwrap().is_empty());
}

#[test]
fn extracts_a_single_belief_clause() {
    let tom = BeliefTrackerTheoryOfMind::new();
    let e = tom.estimate("bob", "bob thinks the meeting is cancelled.");
    let map = beliefs_of(&e);
    let obj = map.as_object().unwrap();
    assert_eq!(obj.len(), 1);
    // Key is "verb:claim"; claim is trimmed and stops at the period.
    let (k, v) = obj.iter().next().unwrap();
    assert_eq!(k, "thinks:the meeting is cancelled");
    // idx 0 ⇒ decay 1.0, non-believe verb ⇒ weight 0.7.
    assert!((v.as_f64().unwrap() - 0.7).abs() < 1e-9);
}

#[test]
fn believe_verbs_weigh_more_than_other_verbs() {
    let tom = BeliefTrackerTheoryOfMind::new();
    // A single believe-clause at idx 0 ⇒ weight 1.0 * decay 1.0 = 1.0.
    let e = tom.estimate("carol", "carol believes the plan will work.");
    let map = beliefs_of(&e);
    let (_, v) = map.as_object().unwrap().iter().next().unwrap();
    assert!((v.as_f64().unwrap() - 1.0).abs() < 1e-9);
}

#[test]
fn later_matches_decay() {
    let tom = BeliefTrackerTheoryOfMind::new();
    // Two non-believe clauses. First idx 0 (decay 1.0), second idx 1 (decay 1/1.1).
    let e = tom.estimate(
        "dave",
        "dave wants a raise; dave fears the layoffs.",
    );
    let map = beliefs_of(&e);
    let obj = map.as_object().unwrap();
    assert_eq!(obj.len(), 2);
    let wants = obj.get("wants:a raise").and_then(|v| v.as_f64()).unwrap();
    let fears = obj.get("fears:the layoffs").and_then(|v| v.as_f64()).unwrap();
    // wants = 0.7 * 1.0 ; fears = 0.7 * (1/1.1) → strictly smaller.
    assert!((wants - 0.7).abs() < 1e-9);
    assert!((fears - 0.7 / 1.1).abs() < 1e-9);
    assert!(fears < wants);
}

#[test]
fn confidence_is_sum_over_five_capped_at_one() {
    let tom = BeliefTrackerTheoryOfMind::new();
    let e = tom.estimate("erin", "erin thinks it is fine.");
    // One clause weight 0.7 ⇒ conf 0.7/5 = 0.14.
    assert!((e.confidence - 0.7 / 5.0).abs() < 1e-9);
}

#[test]
fn confidence_saturates_at_one() {
    let tom = BeliefTrackerTheoryOfMind::new();
    // Many distinct believe clauses drive the summed weight well past 5.
    let mut history = String::new();
    for i in 0..12 {
        history.push_str(&format!("frank believes claim number {i} is true. "));
    }
    let e = tom.estimate("frank", &history);
    assert!(e.confidence <= 1.0);
    assert!((e.confidence - 1.0).abs() < 1e-9, "expected saturation, got {}", e.confidence);
}

#[test]
fn matching_is_case_insensitive() {
    let tom = BeliefTrackerTheoryOfMind::new();
    let e = tom.estimate("gina", "Gina THINKS the door is open.");
    let map = beliefs_of(&e);
    let obj = map.as_object().unwrap();
    assert_eq!(obj.len(), 1);
    // Verb is lower-cased in the key.
    assert!(obj.contains_key("thinks:the door is open"));
}

#[test]
fn stops_the_claim_at_terminators() {
    let tom = BeliefTrackerTheoryOfMind::new();
    // The claim capture is [^.;!?]+ — it must not swallow the exclamation.
    let e = tom.estimate("hank", "hank hopes we win! then we celebrate.");
    let map = beliefs_of(&e);
    assert!(map.as_object().unwrap().contains_key("hopes:we win"));
}

#[test]
#[should_panic(expected = "target required")]
fn rejects_blank_target() {
    let tom = BeliefTrackerTheoryOfMind::new();
    tom.estimate("   ", "anything");
}
